### PR TITLE
Migrate test framework to Hardhat

### PR DESCRIPTION
**Pull request by kronosapiens**
Initial drop-in migration using an external RPC node for testing

- `contracts-network` -- PASSING ✅ 
- `cross-chain` -- PASSING ✅ 
- `extensions` -- PASSING ✅ 
- `packages` -- PASSING ✅
- `reputation-system` -- PASSING ✅ 

- `npx hardhat coverage` -- PASSING ✅
- `REPORT_GAS=true npx hardhat test` -- PASSING ✅

TO DO:

- Migrate smoke tests (see https://github.com/NomicFoundation/hardhat/issues/3345)

NOTES:

- Both `provision:token:contracts` and `provision:safe:contracts` have been incorporated into the compile task.

### CODE REVIEW

**Comment by kronosapiens** [comment]
Noting that the latest version of hardhat (2.21) introduced a new dependency: @nomicfoundation/edr which seems to break our build. So pinning to version 2.19 with a tilde for now.

**Comment by kronosapiens** [comment]
Noting also that the version check script may need to be migrated over. It is passing now but this may be a false negative.

File: test/cross-chain/cross-chain.js [hunk 1]
```diff
@@ -38,43 +49,31 @@ contract("Cross-chain", () => {
...
+        process.exit(1);
       }
+    } else {
+      etherRouterAddress = (await EtherRouter.deployed()).address;
     }
   });
 
   beforeEach(async () => {
     // Set up a colony on the home chain. That may or may not be the truffle chain...
-    // Get the etherrouter address
```
Comment (kronosapiens) [inline test/cross-chain/cross-chain.js:62]: I don't actually remember why this was removed

**Review by area — CHANGES_REQUESTED** [review]
Looking really good, overall!

* You have `cross-chain` listed as "Not running" in the PR, is that a mistake, or am I missing something?
* Is there something fundamentally broken with the version checking (as is the case for the smoke tests), or is it just something you've not got to?

File: package.json [hunk 2]
```diff
@@ -32,46 +35,35 @@
...
-    "test:contracts:patricia": "npm run start:blockchain:client parity & truffle migrate --reset --compile-all && truffle test packages/reputation-miner/patricia-test.js --network development",
-    "test:contracts:coverage": "SOLIDITY_COVERAGE=1 truffle run coverage --network coverage --temp build-coverage --file='./test/contracts-network/*'",
-    "test:contracts:extensions:coverage": "SOLIDITY_COVERAGE=1 truffle run coverage --network coverage --solcoverjs ./.solcover.extensions.js --temp build-coverage --file='./test/extensions/*'",
-    "test:contracts:watch": "npm run start:blockchain:client & truffle migrate --reset --compile-all && truffle watch --network development",
-    "test:contracts:e2e": "npm run start:blockchain:client & truffle migrate --reset --compile-all && truffle test test-system/end-to-end.js --network development",
-    "test:contracts:smoke": "truffle compile && truffle test ./test-smoke/colony-storage-consistent.js --network storageSmoke",
+    "test:reputation:1": "npm run start:blockchain:client && nyc npx hardhat test ./test/reputation-system/*.js --network development",
```
[the remark this answers is not in the document]
Comment (kronosapiens) [reply package.json:38]: Not sure. The only two times we reference `nyc` is `test:reputation:1` and `test:reputation:2`.

File: packages/kyc-oracle/KycOracle.js [hunk 3]
```diff
@@ -168,7 +168,8 @@ class KycOracle {
     const network = await this.provider.getNetwork();
     this.chainId = network.chainId;
 
-    this.whitelistContractDef = await this.loader.load({ contractName: "Whitelist" }, { abi: true, address: false });
+    const flags = { abi: true, address: false };
```
[the remark this answers is not in the document]
Comment (kronosapiens) [reply packages/kyc-oracle/KycOracle.js:171]: Seems like they already are

https://github.com/JoinColony/colonyNetwork/blob/develop/packages/package-utils/TruffleLoader.js#L5

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (kronosapiens) [reply package.json:38]: Might we be able to ditch `nyc` entirely?

File: contracts/testHelpers/TestExtensions.sol [hunk 4]
```diff
@@ -1,103 +0,0 @@
-// SPDX-License-Identifier: GPL-3.0-or-later
```
[the remark this answers is not in the document]
Comment (kronosapiens) [reply contracts/testHelpers/TestExtensions.sol:1]: I'm pretty sure it has to do with having to pass the test extensions through `setupEtherRouter`, which makes some assumptions about contract paths. The `ColonyNetworkAuction` contract contains two contracts, but the second one (`DutchAuction`) is not an `EtherRouter` and so those assumptions are not relevant.

**Comment by kronosapiens** [comment]
Regarding the version check script, I'm not 100% sure. Given that the "last release" will run on truffle and the "current" version will run on hardhat we may need to update that file a few times over the next release cycle to keep the commands / paths accurate. I can keep working on it, but I'm also happy just saying that it won't work for releases which span truffle/hardhat.

**Comment by area** [comment]
Consider this approval, but with provisos:

* I'd like #1227 to be reviewed and merged first, as it's the sort of PR that the storage tests (disabled here) are intended to give confidence in.
* Not a fan of the renamed `TRUFFLE_FOREIGN`. The name was intended to capture whether or not the chain being denominated the foreign chain was the chain Truffle had deployed to. `FOREIGN_CHAIN` to my ear is just whether there is a foreign chain or not, which is always the case when the relevant tests are run. Was there a reason you didn't just end up with `HARDHAT_FOREIGN`?
* I'm not super keen on being pinned to a old version of hardhat. I tried updating (which I agree didn't work out of the box), and installed the package that was missing explicitly, and it worked, but the package that was missing was linux-specific, and we need this to work on OSX as well. Maybe a set of optional dependencies in `package.json` would make that work? It's tough to tell. But we do need to figure it out, as otherwise we'll be back where we are currently with truffle and unable to test against newer forks locally. Let's not hold this up because of it, but do it in another PR imminently.

**Comment by area** [comment]
The tests        `should correctly respond to a request for a reputation state in previous states` and         `should correctly respond to a request for a reputation state in previous states with no proof` seem to have become spotty in this PR. It doesn't seem to be the case in CI, but is definitely the case locally.

Is this only for me, or do you see it as well?

**Comment by kronosapiens** [comment]
@area it's pretty spotty in CI for me, re-running the build usually sorts it. I've been ignoring it for the time being.

**Comment by area** [comment]
I've fixed them, though I'm not totally sold on my fix. What was happening was that when we were explicitly calling `client._miner.sync()` in those tests, the client was (sometimes) still in the process of handling blocks that it had seen, and the two sync processes would clobber each other. I'm pretty sure this is only behaviour that manifests in the tests - the miner is very gun-shy about calling `.sync` itself, and we've written it in a way that shouldn't cause this.

To resolve, I've moved the initialisation of the client out of the `beforeEach` and into the tests themselves. That means the block handler isn't active while we're doing the setup of the state we want. We then call `initialise`, which includes a `sync`. 

However, this is only treating the symptoms. I don't know why this PR cause this behaviour to manifest itself. My best guess is that the specific queries that the client is doing to sync take longer than they did with the Truffle RPC endpoint, and so by the time we hit the `sync` call on the original tests it sometimes hasn't finished. 

Are there any other newly-spotty tests you've noticed?

**Comment by kronosapiens** [comment]
@area that was the only one that was consistently failing

**Pull request merged**